docs/api_requests/block-vhost-user.md

```diff
@@ -0,0 +1,189 @@
+1. [SPDK backend](https://github.com/spdk/spdk/blob/master/lib/vhost/vhost_blk.c)
+
+## Security considerations
+
+### Guest memory sharing
+
+By design, a vhost-user frontend must share file descriptors
+of all guest memory regions to the backend. In order to achive that,
+guest memory is created as a [memfd](https://man7.org/linux/man-pages/man2/memfd_create.2.html)
+and mapped as `MAP_SHARED`.
+
+#### File descriptor in procfs
+
+An open `memfd` is reflected in `procfs` as any other open file descriptor:
+
+```shell
+$ ls -l /proc/{pid}/fd | grep memfd
+lrwx------ 1 1234 1234 64 Nov  2 13:39 32 -> /memfd:guest_mem (deleted)
+```
+
+Any process on the host that has access to this file in `procfs` will be able
+to map the file descriptor and observe runtime behaviour of the guest.
+
+At the moment, Firecracker does not close the `memfd`, because it must remain open
+until all the configured vhost-user devices have been activated and their info
+shared with the backends. This kind of tracking is not implemented in Firecracker,
+but may be implemented in the future. Meanwhile, users need to make sure that
+the access to the Firecracker's `procfs` tree is restricted to trusted processes
+on the host.
+
+On the backend side, it is advised that the backend closes
+the guest memory region file descriptors after mapping them into the address
+space.
+
+#### Resource limit in jailer
+
+The Firecracker [jailer](../jailer.md) allows to configure resource limits
+for the Firecracker process. Specifically, it allows to set the maximum file
+size. Since `memfd` that is used to back the guest memory is considered a file,
+the file size resource limit cannot be less than the biggest guest memory
+region. This does not require any special action from a user, but needs to be
+taken into consideration.
+
+### Remote code execution in the backend
+
+It is recommended to run Firecracker using the [jailer](../jailer.md). Since
+the vhost-user backend interacts with the guest via a Virtio queue, there is
+a potential for the guest to exercise issues in the backend code base
+to trigger undesired behaviours. Users should consider running their backend
+in a jailer or applying other adequate security measures to restrict it.
+
+It is also recommended to use proactive security measures like running
+a Virtio-level fuzzer in the guest during testing to make sure that the backend
+correctly handles all possible classes of inputs (including invalid ones)
+from the guest.
+
+### Rate limiting / cgroups
+
+Virtio block device in Firecracker has a [rate limiting capability](../design.md#io-storage-networking-and-rate-limiting).
+
+In the vhost-user case, Firecracker does not participate in handling requests
+from the guest, so rate limiting becomes backend's responsibility.
+
+As an additional indirect measure, users can make use of `cgroups` settings
+(either via Firecracker jailer or independently) in order to restrict host CPU
```

Comment:
I thought that Firecracker jailer cannot be used to run arbitrary binaries (see previous comment as well). Am I wrong?

Reply:
this is talking about running the guest in the jailer to restrict its CPU time to indirectly influence the amount of IO

Reply:
oh right, I see now! thanks for the clarification